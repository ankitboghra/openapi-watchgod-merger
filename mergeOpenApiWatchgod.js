/*
DEPOSIT
=> priority of responses(highest to lowest)
- openapiDeposits
- watchgodDeposits
*/

/*
WITHDRAW

=> priority of responses(highest to lowest)
- openapiExits
- openapiBurns
- watchgodExits
- watchgodBurns
*/

// interface is just for reference
/*
interface ITransaction {
    from?: String
    to?: String
    amount?: String
    userAddress?: String

    txType?: String
    txStatus?: String
    contractAddress?: String
    rootToken?: String

    network?: String

    tokenId?: String
    tokenType: String
    isPoS?: Boolean

    txHash?: String
    prevBurnHash?: String
    oldHash?: String
    newHash?: String

    data?: String
    blockNumber?: Number

    timestamp?: Number // in seconds
}
*/

const {
    convertOpenapiToInterface,
    convertWatchgodToInterface,
    convertInterfaceToWatchgod,
} = require("./utils");
/**
 *
 * @param {*} openapiDepositResponse
 * @param {*} openapiWithdrawResponse
 * @param {*} openapiBurnResponse
 * @param {*} watchGodResponse
 */
function merger(
    openapiDepositResponse,
    openapiWithdrawResponse,
    openapiBurnResponse,
    watchGodResponse
) {
    const openapiDeposits = openapiDepositResponse.map((tx) =>
        convertOpenapiToInterface(tx)
    );

    const watchgodDeposits = watchGodResponse
        .filter((tx) => tx.txType === "deposit")
        .map((tx) => convertWatchgodToInterface(tx));

    const openapiExits = openapiWithdrawResponse.map((tx) =>
        convertOpenapiToInterface(tx)
    );

    const openapiBurns = openapiBurnResponse.map((tx) =>
        convertOpenapiToInterface(tx)
    );

    const watchgodExits = watchGodResponse
        .filter((tx) => tx.txType === "exit")
        .map((tx) => convertWatchgodToInterface(tx));

    const watchgodBurns = watchGodResponse
        .filter((tx) => tx.txType === "withdraw")
        .map((tx) => convertWatchgodToInterface(tx));

    return [
        ...mergeDeposits(openapiDeposits, watchgodDeposits),
        ...mergeWithdraws(openapiExits, openapiBurns, watchgodExits, watchgodBurns),
    ];
}

function mergeDeposits(openapiDeposits, watchgodDeposits) {
    const mergedDeposits = {};

    watchgodDeposits.forEach((tx) => {
        const txHash = tx.txHash;
        mergedDeposits[txHash] = tx;
        mergedDeposits[txHash].latestStatus = tx.txStatus;
        mergedDeposits[txHash].watchgodTxStatus = tx.txStatus;
        mergedDeposits[txHash].txSource = "watchgod_deposits";
    });

    openapiDeposits.forEach((tx) => {
        const txHash = tx.txHash;
        const watchgodTxStatus = mergedDeposits[txHash]
            ? mergedDeposits[txHash].txStatus
            : false;

        mergedDeposits[txHash] = tx;
        mergedDeposits[txHash].latestStatus = tx.txStatus;
        mergedDeposits[txHash].openapiTxStatus = tx.txStatus;

        if (watchgodTxStatus) {
            mergedDeposits[txHash].watchgodTxStatus = watchgodTxStatus;
        }

        mergedDeposits[txHash].txSource = "openapi";
    });

    return Object.values(mergedDeposits);
}

function mergeWithdraws(
    openapiExits,
    openapiBurns,
    watchgodExits,
    watchgodBurns
) {
    const mergedWithdraws = {};

    watchgodBurns.forEach((tx) => {
        const txHash = tx.txHash;

        mergedWithdraws[txHash] = tx;
        mergedWithdraws[txHash]._txSource = "watchgod_burns";
        mergedWithdraws[txHash]._burnTxHash = tx.txHash;
        mergedWithdraws[txHash]._watchgodBurnTxStatus = tx.txStatus;
        mergedWithdraws[txHash]._latestStatus = tx.txStatus;
    });

    openapiBurns.forEach((tx) => {
        const txHash = tx.txHash;
        const watchgodBurnTxStatus =
            mergedWithdraws[txHash] && mergedWithdraws[txHash]._watchgodBurnTxStatus
                ? mergedWithdraws[txHash]._watchgodBurnTxStatus
                : false;

        mergedWithdraws[txHash] = tx;
        mergedWithdraws[txHash]._txSource = "openapi_burns";
        mergedWithdraws[txHash]._burnTxHash = tx.txHash;
        mergedWithdraws[txHash]._openapiBurnTxStatus = tx.txStatus;
        mergedWithdraws[txHash]._latestStatus = tx.txStatus;

        if (watchgodBurnTxStatus) {
            mergedWithdraws[txHash]._watchgodBurnTxStatus = watchgodBurnTxStatus;
        }
    });

    watchgodExits.forEach((tx) => {
        const txHash = tx.txBurnHash;

        const watchgodBurnTxStatus =
            mergedWithdraws[txHash] && mergedWithdraws[txHash]._watchgodBurnTxStatus
                ? mergedWithdraws[txHash]._watchgodBurnTxStatus
                : false;

        const openapiBurnTxStatus =
            mergedWithdraws[txHash] && mergedWithdraws[txHash]._openapiBurnTxStatus
                ? mergedWithdraws[txHash]._openapiBurnTxStatus
                : false;

        mergedWithdraws[txHash] = tx;
        mergedWithdraws[txHash]._txSource = "watchgod_exits";
        mergedWithdraws[txHash]._burnTxHash = tx.txBurnHash;
        mergedWithdraws[txHash]._exitTxHash = tx.txHash;
        mergedWithdraws[txHash]._watchgodExitTxStatus = tx.txStatus;
        mergedWithdraws[txHash]._latestStatus = tx.txStatus;

        if (watchgodBurnTxStatus) {
            mergedWithdraws[txHash]._watchgodBurnTxStatus = watchgodBurnTxStatus;
        }

        if (openapiBurnTxStatus) {
            mergedWithdraws[txHash]._openapiBurnTxStatus = openapiBurnTxStatus;
        }
    });

    openapiExits.forEach((tx) => {
        const txHash = tx.txBurnHash;

        const watchgodBurnTxStatus =
            mergedWithdraws[txHash] && mergedWithdraws[txHash]._watchgodBurnTxStatus
                ? mergedWithdraws[txHash]._watchgodBurnTxStatus
                : false;

        const openapiBurnTxStatus =
            mergedWithdraws[txHash] && mergedWithdraws[txHash]._openapiBurnTxStatus
                ? mergedWithdraws[txHash]._openapiBurnTxStatus
                : false;

        const watchgodExitTxStatus =
            mergedWithdraws[txHash] && mergedWithdraws[txHash]._watchgodExitTxStatus
                ? mergedWithdraws[txHash]._watchgodExitTxStatus
                : false;

        mergedWithdraws[txHash] = tx;
        mergedWithdraws[txHash]._txSource = "openapi_exits";
        mergedWithdraws[txHash]._burnTxHash = tx.txBurnHash;
        mergedWithdraws[txHash]._exitTxHash = tx.txHash;
        mergedWithdraws[txHash]._openapiExitTxStatus = tx.txStatus;
        mergedWithdraws[txHash]._latestStatus = tx.txStatus;

        if (watchgodBurnTxStatus) {
            mergedWithdraws[txHash]._watchgodBurnTxStatus = watchgodBurnTxStatus;
        }

        if (openapiBurnTxStatus) {
            mergedWithdraws[txHash]._openapiBurnTxStatus = openapiBurnTxStatus;
        }

        if (watchgodExitTxStatus) {
            mergedWithdraws[txHash]._watchgodExitTxStatus = watchgodExitTxStatus;
        }
    });

    return Object.values(mergedWithdraws);
}

module.exports = merger;
