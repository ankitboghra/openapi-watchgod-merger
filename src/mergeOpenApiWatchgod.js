/*
DEPOSIT
=> priority of responses(highest to lowest)
- openapiDeposits
- watchgodDeposits
*/

/*
WITHDRAW

=> priority of responses(highest to lowest)
- openapiExits
- openapiBurns
- watchgodExits
- watchgodBurns
*/

// interface is just for reference
/*
interface ITransaction {
    from?: String
    to?: String
    amount?: String
    userAddress?: String

    txType?: String
    txStatus?: String
    contractAddress?: String
    rootToken?: String

    network?: String

    tokenId?: String
    tokenType: String
    isPoS?: Boolean

    txHash?: String
    prevBurnHash?: String
    oldHash?: String
    newHash?: String

    data?: String
    blockNumber?: Number

    timestamp?: Number // in seconds
}
*/

import {
    convertOpenapiToInterface,
    convertWatchgodToInterface,
    convertInterfaceToWatchgod,
} from "./utils";
import { TX_TYPE } from './constants'


import mergeDeposits from './mergeDeposits'
import mergeWithdraws from './mergeWithdraws'

/**
 *
 * @param {*} openapiDepositResponse
 * @param {*} openapiWithdrawResponse
 * @param {*} openapiBurnResponse
 * @param {*} watchGodResponse
 */
export default function merger(
    openapiDepositResponse,
    openapiWithdrawResponse,
    openapiBurnResponse,
    watchGodResponse
) {
    const openapiDeposits = openapiDepositResponse.map((tx) =>
        convertOpenapiToInterface(tx)
    );

    const watchgodDeposits = watchGodResponse
        .filter((tx) => tx.txType === TX_TYPE.DEPOSIT)
        .map((tx) => convertWatchgodToInterface(tx));

    const openapiExits = openapiWithdrawResponse.map((tx) =>
        convertOpenapiToInterface(tx)
    );

    const openapiBurns = openapiBurnResponse.map((tx) =>
        convertOpenapiToInterface(tx)
    );

    const watchgodExits = watchGodResponse
        .filter((tx) => tx.txType === TX_TYPE.EXIT)
        .map((tx) => convertWatchgodToInterface(tx));

    const watchgodConfirmExits = watchGodResponse
        .filter((tx) => tx.txType === TX_TYPE.CONFIRM_EXIT)
        .map((tx) => convertWatchgodToInterface(tx));

    const watchgodBurns = watchGodResponse
        .filter((tx) => tx.txType === TX_TYPE.WITHDRAW)
        .map((tx) => convertWatchgodToInterface(tx));

    return [
        ...mergeDeposits(openapiDeposits, watchgodDeposits),
        ...mergeWithdraws(openapiExits, openapiBurns, watchgodExits, watchgodConfirmExits, watchgodBurns),
    ];
}
