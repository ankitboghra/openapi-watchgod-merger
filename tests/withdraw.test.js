import mergeOpenApiWatchgod from '../src/index.js'

const watchgodBurn_1 = {
    hash: "0xburnTxHash_1",
    network: "matic-main",
    amount: "13000",
    prevBurnHash: null,
    rootToken: "0xrootToken",
    status: "confirmed",
    txType: "withdraw",
    blockNumber: 21817279,
    data: "0xdata",
    from: "0xfrom",
    timestamp: 1637919854103,
    to: "0xto_1",
}

const openapiExit_1 = {
    transactionStatus: "exited",
    transactionHash: "0xexitHash_1",
    burnTransactionHash: "0xburnTxHash_1",
    userAddress: "0xfrom",
    timestamp: "2022-01-04T05:45:12.000Z",
    rootToken: "0xrootToken",
    blockNumber: 13937154,
    counter: 185187,
    tokenType: "ERC20",
    isPos: true,
    amount: "9857406270252014820230",
    childToken: "0xchildToken",
    burnTxTimestamp: "2022-01-04T04:46:35.000Z",
}

const openapiBurn_1 = {
    transactionHash:
        "0xburnTxHash_1",
    userAddress: "0xfrom",
    childToken: "0xchildToken",
    amount: "13000",
    isPos: true,
    blockNumber: 21817279,
    timestamp: "2021-11-26T09:44:13.000Z",
    transactionStatus: "checkpointed",
    tokenType: "ERC20",
}

const watchgodExit_1 = {
    hash: "0xexitHash_1",
    network: "main",
    amount: "13000",
    prevBurnHash: "0xburnTxHash_1",
    rootToken: "0xrootToken",
    status: "confirmed",
    txType: "exit",
    blockNumber: 13689313,
    data: "0xdata",
    from: "0xfrom",
    timestamp: 1637924277824,
    to: "0xto_1",
}



test('watchgod burn only', () => {
    mergeOpenApiWatchgod([], [], [], [watchgodBurn_1]).forEach((tx) => {
        expect(tx).toMatchSnapshot()
    })
});

test('openapi burn only', () => {
    mergeOpenApiWatchgod([], [], [openapiBurn_1], []).forEach((tx) => {
        expect(tx).toMatchSnapshot()
        // expect(tx).toHaveProperty('_latestStatus', 'confirmed')
        // expect(tx).toHaveProperty('_watchgodTxStatus', 'confirmed')
        // expect(tx).toHaveProperty('_txSource', 'watchgod_deposits')
    })
});

test('watchgod exit only', () => {
    mergeOpenApiWatchgod([], [], [], [watchgodExit_1]).forEach((tx) => {
        expect(tx).toMatchSnapshot()
        // expect(tx).toHaveProperty('_latestStatus', 'confirmed')
        // expect(tx).toHaveProperty('_watchgodTxStatus', 'confirmed')
        // expect(tx).toHaveProperty('_txSource', 'watchgod_deposits')
    })
});

test('openapi exit only', () => {
    mergeOpenApiWatchgod([], [openapiExit_1], [], []).forEach((tx) => {
        expect(tx).toMatchSnapshot()
        // expect(tx).toHaveProperty('_latestStatus', 'confirmed')
        // expect(tx).toHaveProperty('_watchgodTxStatus', 'confirmed')
        // expect(tx).toHaveProperty('_txSource', 'watchgod_deposits')
    })
});

test('watchgod burn and openapi burn', () => {
    mergeOpenApiWatchgod([], [], [openapiBurn_1], [watchgodBurn_1]).forEach((tx) => {
        expect(tx).toMatchSnapshot()
        // expect(tx).toHaveProperty('_latestStatus', 'confirmed')
        // expect(tx).toHaveProperty('_watchgodTxStatus', 'confirmed')
        // expect(tx).toHaveProperty('_txSource', 'watchgod_deposits')
    })
});

test('watchgod burn, openapi burn, watchgod exit', () => {
    mergeOpenApiWatchgod([], [], [openapiBurn_1], [watchgodBurn_1, watchgodExit_1]).forEach((tx) => {
        expect(tx).toMatchSnapshot()
        // expect(tx).toHaveProperty('_latestStatus', 'confirmed')
        // expect(tx).toHaveProperty('_watchgodTxStatus', 'confirmed')
        // expect(tx).toHaveProperty('_txSource', 'watchgod_deposits')
    })
});

test('watchgod burn, openapi burn, watchgod exit, openapi exit', () => {
    mergeOpenApiWatchgod([], [openapiExit_1], [openapiBurn_1], [watchgodBurn_1, watchgodExit_1]).forEach((tx) => {
        expect(tx).toMatchSnapshot()
        // expect(tx).toHaveProperty('_latestStatus', 'confirmed')
        // expect(tx).toHaveProperty('_watchgodTxStatus', 'confirmed')
        // expect(tx).toHaveProperty('_txSource', 'watchgod_deposits')
    })
});

test('openapi burn, openapi exit', () => {
    mergeOpenApiWatchgod([], [openapiExit_1], [openapiBurn_1], []).forEach((tx) => {
        expect(tx).toMatchSnapshot()
        // expect(tx).toHaveProperty('_latestStatus', 'confirmed')
        // expect(tx).toHaveProperty('_watchgodTxStatus', 'confirmed')
        // expect(tx).toHaveProperty('_txSource', 'watchgod_deposits')
    })
});

test('watchgod burn, openapi exit', () => {
    mergeOpenApiWatchgod([], [openapiExit_1], [], [watchgodBurn_1]).forEach((tx) => {
        expect(tx).toMatchSnapshot()
        // expect(tx).toHaveProperty('_latestStatus', 'confirmed')
        // expect(tx).toHaveProperty('_watchgodTxStatus', 'confirmed')
        // expect(tx).toHaveProperty('_txSource', 'watchgod_deposits')
    })
});
