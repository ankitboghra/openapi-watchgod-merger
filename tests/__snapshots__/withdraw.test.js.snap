// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`openapi burn only 1`] = `
Object {
  "_burnTxHash": "0xburnTxHash_1",
  "_latestStatus": "checkpointed",
  "_openapiBurnTxStatus": "checkpointed",
  "_txSource": "openapi_burns",
  "_txType": "withdraw",
  "amount": "13000",
  "blockNumber": 21817279,
  "childToken": "0xchildToken",
  "isPoS": true,
  "timestamp": 1637919853,
  "tokenType": "ERC20",
  "txHash": "0xburnTxHash_1",
  "txStatus": "checkpointed",
  "userAddress": "0xfrom",
}
`;

exports[`openapi burn, openapi exit 1`] = `
Object {
  "_burnTxHash": "0xburnTxHash_1",
  "_exitTxHash": "0xexitHash_1",
  "_latestStatus": "exited",
  "_openapiBurnTxStatus": "checkpointed",
  "_openapiExitTxStatus": "exited",
  "_txSource": "openapi_exits",
  "_txType": "exit",
  "amount": "9857406270252014820230",
  "blockNumber": 13937154,
  "burnTxTimestamp": "2022-01-04T04:46:35.000Z",
  "childToken": "0xchildToken",
  "counter": 185187,
  "isPoS": true,
  "rootToken": "0xrootToken",
  "timestamp": 1641275112,
  "tokenType": "ERC20",
  "txBurnHash": "0xburnTxHash_1",
  "txHash": "0xexitHash_1",
  "txStatus": "exited",
  "userAddress": "0xfrom",
}
`;

exports[`openapi exit only 1`] = `
Object {
  "_burnTxHash": "0xburnTxHash_1",
  "_exitTxHash": "0xexitHash_1",
  "_latestStatus": "exited",
  "_openapiExitTxStatus": "exited",
  "_txSource": "openapi_exits",
  "_txType": "exit",
  "amount": "9857406270252014820230",
  "blockNumber": 13937154,
  "burnTxTimestamp": "2022-01-04T04:46:35.000Z",
  "childToken": "0xchildToken",
  "counter": 185187,
  "isPoS": true,
  "rootToken": "0xrootToken",
  "timestamp": 1641275112,
  "tokenType": "ERC20",
  "txBurnHash": "0xburnTxHash_1",
  "txHash": "0xexitHash_1",
  "txStatus": "exited",
  "userAddress": "0xfrom",
}
`;

exports[`watchgod burn and openapi burn 1`] = `
Object {
  "_burnTxHash": "0xburnTxHash_1",
  "_latestStatus": "checkpointed",
  "_openapiBurnTxStatus": "checkpointed",
  "_txSource": "openapi_burns",
  "_txType": "withdraw",
  "_watchgodBurnTxStatus": "confirmed",
  "amount": "13000",
  "blockNumber": 21817279,
  "childToken": "0xchildToken",
  "isPoS": true,
  "timestamp": 1637919853,
  "tokenType": "ERC20",
  "txHash": "0xburnTxHash_1",
  "txStatus": "checkpointed",
  "userAddress": "0xfrom",
}
`;

exports[`watchgod burn only 1`] = `
Object {
  "_burnTxHash": "0xburnTxHash_1",
  "_latestStatus": "confirmed",
  "_txSource": "watchgod_burns",
  "_txType": "withdraw",
  "_watchgodBurnTxStatus": "confirmed",
  "amount": "13000",
  "blockNumber": 21817279,
  "data": "0xdata",
  "from": "0xfrom",
  "network": "matic-main",
  "rootToken": "0xrootToken",
  "timestamp": 1637919854103,
  "to": "0xto_1",
  "txBurnHash": null,
  "txHash": "0xburnTxHash_1",
  "txStatus": "confirmed",
  "txType": "withdraw",
}
`;

exports[`watchgod burn, openapi burn, watchgod exit 1`] = `
Object {
  "_burnTxHash": "0xburnTxHash_1",
  "_exitTxHash": "0xexitHash_1",
  "_latestStatus": "confirmed",
  "_openapiBurnTxStatus": "checkpointed",
  "_txSource": "watchgod_exits",
  "_txType": "exit",
  "_watchgodBurnTxStatus": "confirmed",
  "_watchgodExitTxStatus": "confirmed",
  "amount": "13000",
  "blockNumber": 13689313,
  "data": "0xdata",
  "from": "0xfrom",
  "network": "main",
  "rootToken": "0xrootToken",
  "timestamp": 1637924277824,
  "to": "0xto_1",
  "txBurnHash": "0xburnTxHash_1",
  "txHash": "0xexitHash_1",
  "txStatus": "confirmed",
  "txType": "exit",
}
`;

exports[`watchgod burn, openapi burn, watchgod exit, openapi exit 1`] = `
Object {
  "_burnTxHash": "0xburnTxHash_1",
  "_exitTxHash": "0xexitHash_1",
  "_latestStatus": "exited",
  "_openapiBurnTxStatus": "checkpointed",
  "_openapiExitTxStatus": "exited",
  "_txSource": "openapi_exits",
  "_txType": "exit",
  "_watchgodBurnTxStatus": "confirmed",
  "_watchgodExitTxStatus": "confirmed",
  "amount": "9857406270252014820230",
  "blockNumber": 13937154,
  "burnTxTimestamp": "2022-01-04T04:46:35.000Z",
  "childToken": "0xchildToken",
  "counter": 185187,
  "isPoS": true,
  "rootToken": "0xrootToken",
  "timestamp": 1641275112,
  "tokenType": "ERC20",
  "txBurnHash": "0xburnTxHash_1",
  "txHash": "0xexitHash_1",
  "txStatus": "exited",
  "userAddress": "0xfrom",
}
`;

exports[`watchgod burn, openapi exit 1`] = `
Object {
  "_burnTxHash": "0xburnTxHash_1",
  "_exitTxHash": "0xexitHash_1",
  "_latestStatus": "exited",
  "_openapiExitTxStatus": "exited",
  "_txSource": "openapi_exits",
  "_txType": "exit",
  "_watchgodBurnTxStatus": "confirmed",
  "amount": "9857406270252014820230",
  "blockNumber": 13937154,
  "burnTxTimestamp": "2022-01-04T04:46:35.000Z",
  "childToken": "0xchildToken",
  "counter": 185187,
  "isPoS": true,
  "rootToken": "0xrootToken",
  "timestamp": 1641275112,
  "tokenType": "ERC20",
  "txBurnHash": "0xburnTxHash_1",
  "txHash": "0xexitHash_1",
  "txStatus": "exited",
  "userAddress": "0xfrom",
}
`;

exports[`watchgod exit only 1`] = `
Object {
  "_burnTxHash": "0xburnTxHash_1",
  "_exitTxHash": "0xexitHash_1",
  "_latestStatus": "confirmed",
  "_txSource": "watchgod_exits",
  "_txType": "exit",
  "_watchgodExitTxStatus": "confirmed",
  "amount": "13000",
  "blockNumber": 13689313,
  "data": "0xdata",
  "from": "0xfrom",
  "network": "main",
  "rootToken": "0xrootToken",
  "timestamp": 1637924277824,
  "to": "0xto_1",
  "txBurnHash": "0xburnTxHash_1",
  "txHash": "0xexitHash_1",
  "txStatus": "confirmed",
  "txType": "exit",
}
`;
